plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/TransactionLogAccess.java

```diff
@@ -283,31 +284,33 @@ public static ImmutableList<DeltaLakeColumnMetadata> columnsWithStats(List<Delta
                 .collect(toImmutableList());
     }
 
-    private Stream<AddFileEntry> activeAddEntries(Stream<DeltaLakeTransactionLogEntry> checkpointEntries, Stream<DeltaLakeTransactionLogEntry> jsonEntries)
+    private Stream<AddFileEntry> activeAddEntries(Stream<DeltaLakeTransactionLogEntry> checkpointEntries, Map<Long, List<DeltaLakeTransactionLogEntry>> jsonEntries)
     {
         Map<String, AddFileEntry> activeJsonEntries = new LinkedHashMap<>();
         HashSet<String> removedFiles = new HashSet<>();
-        HashSet<String> dependOnDeletionVector = new HashSet<>();
 
         // The json entries containing the last few entries in the log need to be applied on top of the parquet snapshot:
         // - Any files which have been removed need to be excluded
         // - Any files with newer add actions need to be updated with the most recent metadata
-        jsonEntries.forEach(deltaLakeTransactionLogEntry -> {
-            AddFileEntry addEntry = deltaLakeTransactionLogEntry.getAdd();
-            if (addEntry != null) {
-                activeJsonEntries.put(addEntry.getPath(), addEntry);
-                addEntry.getDeletionVector().ifPresent(deletionVector -> dependOnDeletionVector.add(addEntry.getPath()));
-            }
+        for (Map.Entry<Long, List<DeltaLakeTransactionLogEntry>> deltaLakeTransactionLogEntries : jsonEntries.entrySet()) {
+            // Deletion vector registers both 'add' & 'remove' entries in any order. The 'add' entry should be kept.
+            Set<String> dependOnDeletionVector = new HashSet<>();
```

Comment:
if we process removals before additions, can we remove `dependOnDeletionVector` set?

Reply:
Updated to process removals first.